rfq/manager.go

```diff
@@ -0,0 +1,583 @@
+		//  accept message corresponds to a request.
+		//
+		// The quote request has been accepted. Store accepted quote
+		// so that it can be used to send a payment by our lightning
+		// node.
+		m.peerAcceptedQuotesMtx.Lock()
+		defer m.peerAcceptedQuotesMtx.Unlock()
+
+		scid := SerialisedScid(msg.ShortChannelId())
+		m.peerAcceptedQuotes[scid] = *msg
+
+		// Notify subscribers of the incoming quote accept.
+		event := NewIncomingAcceptQuoteEvent(msg)
+		m.publishSubscriberEvent(event)
+
+	case *rfqmsg.Reject:
+		// The quote request has been rejected. Notify subscribers of
+		// the rejection.
+		event := NewIncomingRejectQuoteEvent(msg)
+		m.publishSubscriberEvent(event)
+	}
+
+	return nil
+}
+
+// handleOutgoingMessage handles an outgoing message. Outgoing messages are
+// messages that will be sent to a peer.
+func (m *Manager) handleOutgoingMessage(outgoingMsg rfqmsg.OutgoingMsg) error {
+	// Perform type specific handling of the outgoing message.
+	switch msg := outgoingMsg.(type) {
+	case *rfqmsg.Accept:
+		// Before sending an accept message to a peer, inform the HTLC
+		// order handler that we've accepted the quote request.
+		err := m.orderHandler.RegisterChannelRemit(*msg)
+		if err != nil {
+			return fmt.Errorf("error registering channel remit: %w",
+				err)
+		}
+	}
+
+	// Send the outgoing message to the peer.
+	err := m.streamHandler.HandleOutgoingMessage(outgoingMsg)
+	if err != nil {
+		return fmt.Errorf("error sending outgoing message to stream "+
+			"handler: %w", err)
+	}
+
+	return nil
+}
+
+// mainEventLoop is the main event loop of the RFQ manager.
+func (m *Manager) mainEventLoop() {
+	for {
+		select {
+		// Handle incoming message.
+		case incomingMsg := <-m.incomingMessages:
+			peer := incomingMsg.MsgPeer()
+			log.Debugf("Manager handling incoming message "+
+				"(msg_type=%T, origin_peer=%s)", incomingMsg,
+				peer)
+
+			err := m.handleIncomingMessage(incomingMsg)
+			if err != nil {
+				log.Warnf("Error handling incoming message: %v",
+					err)
+			}
+
+		// Handle outgoing message.
+		case outgoingMsg := <-m.outgoingMessages:
+			peer := outgoingMsg.MsgPeer()
+			log.Debugf("Manager handling outgoing message "+
+				"(msg_type=%T, dest_peer=%s)", outgoingMsg,
+				peer.String())
+
+			err := m.handleOutgoingMessage(outgoingMsg)
+			if err != nil {
+				log.Warnf("Error handling outgoing message: %v",
+					err)
+			}
+
+		case acceptHtlcEvent := <-m.acceptHtlcEvents:
+			// Handle a HTLC accept event. Notify any subscribers.
+			m.publishSubscriberEvent(&acceptHtlcEvent)
+
+		// Handle errors from the negotiator.
+		case err := <-m.negotiator.ErrChan:
+			log.Warnf("Negotiator has encountered an error: %v",
+				err)
+
+		case <-m.Quit:
+			log.Debug("Manager main event loop has received the " +
+				"shutdown signal")
+			return
+		}
+	}
+}
+
+// UpsertAssetSellOffer upserts an asset sell offer with the RFQ manager.
+func (m *Manager) UpsertAssetSellOffer(offer SellOffer) error {
```

Comment:
I don't understand you're thinking on the verb "upsert". Do you think that we should only use it when naming SQL statements?

> since we don't know on what parameters of the offer the "collision" (replacement) is happening

On the asset. If a sell offer exists for an asset then its updated. I'll add that in the doc.

Reply:
Yes, I mean the verb "upsert". It's not an actual English word. And anyone unfamiliar with relational databases perhaps won't know the meaning. So I think it's fine to use it within the `tapd` package. But everywhere else I wouldn't use it, especially not in user-facing RPC names.

Reply:
If you think the programmers who will use the RPC endpoints won't understand/be comfortable with "upsert" when trying to create or update a sell offer/buy order then we should rename.

Reply:
In the latest commit I've changed "upsert" to "add" in the RPC endpoints. And I've described the update behaviour in the comment.